File: src/remote-access.ts
import { principalHasGrant, type HostedPrincipal } from "./hosted-principals.js";
import { PROJECT_API_ROUTES } from "./project-api-contract.js";

export type RemoteActorRole = "owner" | "guest" | "operator";

export interface RemoteActor {
  role: RemoteActorRole;
  userId?: string;
  displayName?: string;
  email?: string;
  shareId?: string;
  shareSessionId?: string;
  /**
   * Hosted-mode principal. Server-minted only — `parseRemoteActor` never
   * produces one, because the relay cannot authenticate an operator and an
   * inbound header claiming the role is forgery.
   */
  principal?: HostedPrincipal;
}

export interface RemoteAccessContext {
  /** Parsed request body: POST routes carry `sessionId` there, not in the query. */
  body?: unknown;
  /** Project root owning the proxied service port. Operators are denied without it. */
  projectRoot?: string | null;
}

export interface RemoteAccessDecision {
  ok: boolean;
  status?: number;
  error?: string;
}

const ACTOR_HEADER = "x-aimux-actor";
const ROLE_HEADER = "x-aimux-actor-role";
const USER_ID_HEADER = "x-aimux-actor-user-id";
const DISPLAY_NAME_HEADER = "x-aimux-actor-display-name";
const EMAIL_HEADER = "x-aimux-actor-email";
const SHARE_ID_HEADER = "x-aimux-share-id";
const SHARE_SESSION_ID_HEADER = "x-aimux-share-session-id";
const RELAY_HEADER_PREFIX = "x-aimux-";
const SHARED_GUEST_SESSION_READ_ROUTES = new Set<string>([
  PROJECT_API_ROUTES.agents.output,
  PROJECT_API_ROUTES.agents.history,
  PROJECT_API_ROUTES.livePane.output,
  PROJECT_API_ROUTES.events,
]);
const SHARED_GUEST_SESSION_WRITE_ROUTES = new Set<string>([
  PROJECT_API_ROUTES.livePane.input,
  PROJECT_API_ROUTES.attachments,
]);
const ATTACHMENT_CONTENT_ROUTE = /^\/attachments\/[A-Za-z0-9_-]{1,128}\/content$/;

const PROXY_PATH_PATTERN = /^\/proxy\/[^/]+\/\d+(\/.*)$/;

/**
 * Every route an operator may reach, and the only method each accepts.
 *
 * Deny-by-default: a route added elsewhere in the codebase is unreachable to
 * operators until someone deliberately lists it here. `/agents` (list) is
 * absent on purpose — it returns every session in the project, which would leak
 * other principals' session ids. `/agents/output/stream` is absent because it
 * belongs to `OPERATOR_STREAM_ROUTES` below and must never appear here, and
 * `/events` because it is project-wide rather than per-session.
 */
const OPERATOR_ROUTE_METHODS = new Map<string, "GET" | "POST">([
  [PROJECT_API_ROUTES.agents.output, "GET"],
  [PROJECT_API_ROUTES.agents.input, "POST"],
  // Write-only by construction: the route replaces or clears, and there is no
  // GET, so a principal can steer its own granted session's prompts and cannot
  // read back what any other client put there.
  [PROJECT_API_ROUTES.agents.promptContext, "POST"],
  [PROJECT_API_ROUTES.agents.interrupt, "POST"],
  [PROJECT_API_ROUTES.attachments, "POST"],
]);

/**
 * Routes whose path carries an id, and so cannot be a Map key.
 *
 * Consulted only after the exact allowlist misses, and deliberately tiny: a
 * pattern is a weaker statement than a literal, so anything expressible as a
 * literal stays one. `/attachments/<id>` (metadata) is absent — the client
 * gets everything it needs from the transcript and the upload reply, so there
 * is no reason to widen the surface for it.
 *
 * The character class is doing real work. It admits no `.`, `/`, `%`, `;` or
 * `?`, which is what stops a traversal, an encoded slash, a matrix parameter
 * or a query fragment from riding inside the id and landing on a different
 * route. Ids are `att_` plus 32 hex characters, comfortably inside it. The
 * path reaching here has already been through `new URL()`, so `.` and `..`
 * segments — and their percent-encoded spellings — are gone before matching;
 * the class is the second of the two defences, not the only one.
 */
const OPERATOR_ROUTE_PATTERNS: ReadonlyArray<{ pattern: RegExp; method: "GET" | "POST" }> = [
  { pattern: /^\/attachments\/[A-Za-z0-9_-]{1,128}\/content$/, method: "GET" },
];

/**
 * Streaming routes, kept in a SEPARATE gate reached by a separate function.
 *
 * They must never appear in `OPERATOR_ROUTE_METHODS`: that allowlist governs
 * the buffered proxy, which reads the whole response before replying, and an
 * SSE body never ends. A stream arriving there would hang the request forever.
 *
 * `/events` is absent on purpose — it is project-wide rather than per-session,
 * so it would carry other operators' sessions to whoever subscribed.
 */
const OPERATOR_STREAM_ROUTES = new Set<string>([PROJECT_API_ROUTES.agents.outputStream]);

function trimmedString(value: unknown): string | null {
  return typeof value === "string" && value.trim() ? value.trim() : null;
}

function bodySessionId(body: unknown): string | null {
  if (!body || typeof body !== "object") return null;
  return trimmedString((body as Record<string, unknown>).sessionId);
}

function bodyText(body: unknown): string | null {
  if (!body || typeof body !== "object") return null;
  return trimmedString((body as Record<string, unknown>).text);
}

function bodyHasAttachments(body: unknown): boolean {
  if (!body || typeof body !== "object") return false;
  const attachmentIds = (body as Record<string, unknown>).attachmentIds;
  return Array.isArray(attachmentIds)
    ? attachmentIds.some((id) => (typeof id === "string" ? Boolean(id.trim()) : id != null))
    : false;
}

/**
 * Authorize the session id the SERVICE will actually act on.
 *
 * The project service reads `sessionId` from the body on POST routes and from
 * the query on GET routes. Checking the wrong source — or accepting either —
 * would let a request authorize one session and operate on another, so the
 * authoritative source is chosen per route and a conflicting value in the other
 * one is refused outright rather than ignored.
 */
function authorizedSessionId(
  method: "GET" | "POST",
  searchParams: URLSearchParams,
  body: unknown,
  label = "operator route",
): { ok: true; sessionId: string } | { ok: false; error: string } {
  const fromBody = bodySessionId(body);
  const fromQuery = trimmedString(searchParams.get("sessionId"));
  const authoritative = method === "POST" ? fromBody : fromQuery;
  const other = method === "POST" ? fromQuery : fromBody;

  if (!authoritative) return { ok: false, error: `${label} requires a session id` };
  if (other && other !== authoritative) {
    return { ok: false, error: "conflicting session ids in request body and query" };
  }
  return { ok: true, sessionId: authoritative };
}

function assertSharedGuestSession(
  actor: RemoteActor,
  method: "GET" | "POST",
  searchParams: URLSearchParams,
  context: RemoteAccessContext,
): RemoteAccessDecision {
  if (!actor.shareSessionId) {
    return { ok: false, status: 403, error: "shared guest route requires an authorized share session" };
  }
  const session = authorizedSessionId(method, searchParams, context.body, "shared session route");
  if (!session.ok) return { ok: false, status: 403, error: session.error };
  if (session.sessionId !== actor.shareSessionId) {
    return { ok: false, status: 403, error: "shared guest cannot access another session" };
  }
  return { ok: true };
}

/**
 * The proxied sub-path an operator is asking for, or a refusal.
 *
 * `denied` rather than an `ok` flag: `RemoteAccessDecision` is not a
 * discriminated union, so a success-shaped value returned here would type-check
 * as a decision and skip the grant check entirely.
 */
function operatorSubPath(
  actor: RemoteActor,
  pathname: string,
): { denied: RemoteAccessDecision } | { denied?: undefined; principal: HostedPrincipal; subPath: string } {
  const principal = actor.principal;
  if (!principal) {
    return { denied: { ok: false, status: 403, error: "operator actor is missing its principal" } };
  }

  const proxyMatch = pathname.match(PROXY_PATH_PATTERN);
  if (!proxyMatch) {
    return { denied: { ok: false, status: 403, error: "operators cannot access daemon routes" } };
  }

  return { principal, subPath: proxyMatch[1] || "/" };
}

/** The grant check both gates share: a session id, bound to a project root. */
function assertGrantedSession(
  principal: HostedPrincipal,
  method: "GET" | "POST",
  searchParams: URLSearchParams,
  context: RemoteAccessContext,
): RemoteAccessDecision {
  // Without a project root the grant check would compare session ids alone, and
  // a grant in one project would authorize the same name in another.
  const projectRoot = context.projectRoot;
  if (!projectRoot) return { ok: false, status: 403, error: "operator request could not be bound to a project" };

  const session = authorizedSessionId(method, searchParams, context.body);
  if (!session.ok) return { ok: false, status: 403, error: session.error };

  if (!principalHasGrant(principal, { projectRoot, sessionId: session.sessionId })) {
    return { ok: false, status: 403, error: "operator is not granted this session" };
  }

  return { ok: true };
}

function assertOperatorAllowed(
  actor: RemoteActor,
  method: string,
  pathname: string,
  searchParams: URLSearchParams,
  context: RemoteAccessContext,
): RemoteAccessDecision {
  const resolved = operatorSubPath(actor, pathname);
  if (resolved.denied) return resolved.denied;

  const allowedMethod =
    OPERATOR_ROUTE_METHODS.get(resolved.subPath) ??
    OPERATOR_ROUTE_PATTERNS.find((route) => route.pattern.test(resolved.subPath))?.method;
  if (!allowedMethod) return { ok: false, status: 403, error: "route is not available to operators" };
  if (method.toUpperCase() !== allowedMethod) {
    return { ok: false, status: 403, error: "method not allowed for this route" };
  }

  return assertGrantedSession(resolved.principal, allowedMethod, searchParams, context);
}

/**
 * The streaming gate, deliberately a separate entry point.
 *
 * A boolean on the shared context would fail open: a caller that forgot to set
 * it would silently get the other allowlist. Streams and buffered requests are
 * served by different machinery and must be asked for by name.
 */
export function assertOperatorStreamAllowed(
  actor: RemoteActor | null,
  method: string,
  pathname: string,
  searchParams: URLSearchParams,
  context: RemoteAccessContext = {},
): RemoteAccessDecision {
  // Null denies rather than throwing, and unlike the buffered gate it does NOT
  // mean "local, allow everything": nothing reaches this without a verified
  // token, so a null actor here is a bug, and a bug must fail closed.
  if (actor?.role !== "operator") return { ok: false, status: 403, error: "streaming requires an operator" };
  if (method.toUpperCase() !== "GET") return { ok: false, status: 403, error: "streams are GET only" };

  const resolved = operatorSubPath(actor, pathname);
  if (resolved.denied) return resolved.denied;

  if (!OPERATOR_STREAM_ROUTES.has(resolved.subPath)) {
    return { ok: false, status: 403, error: "route is not available to operators" };
  }

  return assertGrantedSession(resolved.principal, "GET", searchParams, context);
}

function headerValue(headers: Record<string, string> | undefined, name: string): string | undefined {
  if (!headers) return undefined;
  const direct = headers[name];
  if (typeof direct === "string") return direct.trim() || undefined;
  const lowerName = name.toLowerCase();
  for (const [key, value] of Object.entries(headers)) {
    if (key.toLowerCase() === lowerName) return value.trim() || undefined;
  }
  return undefined;
}

function hasRelayActorHeaders(headers: Record<string, string> | undefined): boolean {
  if (!headers) return false;
  return Object.keys(headers).some((key) => key.toLowerCase().startsWith(RELAY_HEADER_PREFIX));
}

function actorFromJson(value: string): Partial<RemoteActor> | null {
  try {
    const parsed = JSON.parse(value) as unknown;
    if (!parsed || typeof parsed !== "object") return null;
    const record = parsed as Record<string, unknown>;
    return {
      role: record.role === "owner" || record.role === "guest" ? record.role : undefined,
      userId: typeof record.userId === "string" ? record.userId : undefined,
      displayName: typeof record.displayName === "string" ? record.displayName : undefined,
      email: typeof record.email === "string" ? record.email : undefined,
    };
  } catch {
    return null;
  }
}

export function parseRemoteActor(headers: Record<string, string> | undefined): RemoteActor | null {
  const actorJson = headerValue(headers, ACTOR_HEADER);
  const jsonActor = actorJson ? actorFromJson(actorJson) : null;
  const role = headerValue(headers, ROLE_HEADER) ?? jsonActor?.role;
  if (!role) {
    return hasRelayActorHeaders(headers) ? { role: "guest" } : null;
  }
  // Only owner and guest are mintable from headers. "operator" in particular
  // must degrade rather than pass: hosted mode constructs operators from a
  // verified bearer token, so a header claiming the role is forgery.
  if (role !== "owner" && role !== "guest") {
    return { role: "guest" };
  }
  return {
    role,
    userId: headerValue(headers, USER_ID_HEADER) ?? jsonActor?.userId,
    displayName: headerValue(headers, DISPLAY_NAME_HEADER) ?? jsonActor?.displayName,
    email: headerValue(headers, EMAIL_HEADER) ?? jsonActor?.email,
    shareId: headerValue(headers, SHARE_ID_HEADER),
    shareSessionId: headerValue(headers, SHARE_SESSION_ID_HEADER),
  };
}

export function assertRemoteAccessAllowed(
  actor: RemoteActor | null,
  method: string,
  pathname: string,
  searchParams: URLSearchParams,
  context: RemoteAccessContext = {},
): RemoteAccessDecision {
  if (!actor || actor.role === "owner") return { ok: true };
  if (actor.role === "operator") return assertOperatorAllowed(actor, method, pathname, searchParams, context);
  if (actor.role !== "guest") return { ok: false, status: 403, error: "remote actor role is not allowed" };
  const methodName = method.toUpperCase();
  if (methodName === "GET" && pathname === "/health") return { ok: true };

  const proxyMatch = pathname.match(PROXY_PATH_PATTERN);
  if (!proxyMatch) return { ok: false, status: 403, error: "shared guests cannot access daemon routes" };

  const subPath = proxyMatch[1] || "/";
  if (
    methodName === "GET" &&
    (SHARED_GUEST_SESSION_READ_ROUTES.has(subPath) || ATTACHMENT_CONTENT_ROUTE.test(subPath))
  ) {
    return assertSharedGuestSession(actor, "GET", searchParams, context);
  }

  if (methodName === "POST" && SHARED_GUEST_SESSION_WRITE_ROUTES.has(subPath)) {
    const sessionDecision = assertSharedGuestSession(actor, "POST", searchParams, context);
    if (!sessionDecision.ok) return sessionDecision;
    if (subPath === PROJECT_API_ROUTES.livePane.input && !bodyText(context.body) && !bodyHasAttachments(context.body)) {
      return { ok: false, status: 403, error: "shared guest input requires text or attachments" };
    }
    return { ok: true };
  }

  if (methodName !== "GET") {
    return { ok: false, status: 403, error: "shared guests can only write to their shared session" };
  }
  return { ok: false, status: 403, error: "shared guests can only read shared session output and attachments" };
}

File: native/crates/aimux/src/remote_access.rs
use crate::daemon::routing::DaemonRouteUrl;
use crate::project_api_contract::routes as project_routes;
use crate::proxy_project_binding::parse_proxy_target;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RemoteActorRole {
    Owner,
    Guest,
    Operator,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RemoteOperatorGrant {
    pub project_root: String,
    pub session_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RemoteOperatorPrincipal {
    pub id: String,
    pub label: String,
    pub role: String,
    pub grants: Vec<RemoteOperatorGrant>,
    pub revoked_at: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RemoteActor {
    pub role: RemoteActorRole,
    pub user_id: Option<String>,
    pub display_name: Option<String>,
    pub email: Option<String>,
    pub share_id: Option<String>,
    pub share_session_id: Option<String>,
    pub principal: Option<RemoteOperatorPrincipal>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct RemoteAccessContext<'a> {
    pub body: Option<&'a Value>,
    pub project_root: Option<&'a str>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RemoteAccessDecision {
    pub ok: bool,
    pub status: Option<u16>,
    pub error: Option<String>,
}

impl RemoteAccessDecision {
    pub fn allow() -> Self {
        Self {
            ok: true,
            status: None,
            error: None,
        }
    }

    pub fn deny(status: u16, error: impl Into<String>) -> Self {
        Self {
            ok: false,
            status: Some(status),
            error: Some(error.into()),
        }
    }
}

pub fn parse_remote_actor(headers: &BTreeMap<String, String>) -> Option<RemoteActor> {
    let actor_json = header_value(headers, "x-aimux-actor");
    let json_actor = actor_json.and_then(actor_from_json);
    let role_text = header_value(headers, "x-aimux-actor-role")
        .or_else(|| json_actor.as_ref().and_then(|actor| actor.role.as_deref()));
    let Some(role_text) = role_text else {
        return has_relay_actor_headers(headers).then_some(RemoteActor {
            role: RemoteActorRole::Guest,
            user_id: None,
            display_name: None,
            email: None,
            share_id: None,
            share_session_id: None,
            principal: None,
        });
    };
    let role = match role_text {
        "owner" => RemoteActorRole::Owner,
        "guest" => RemoteActorRole::Guest,
        _ => RemoteActorRole::Guest,
    };
    Some(RemoteActor {
        role,
        user_id: header_value(headers, "x-aimux-actor-user-id")
            .map(str::to_owned)
            .or_else(|| json_actor.as_ref().and_then(|actor| actor.user_id.clone())),
        display_name: header_value(headers, "x-aimux-actor-display-name")
            .map(str::to_owned)
            .or_else(|| {
                json_actor
                    .as_ref()
                    .and_then(|actor| actor.display_name.clone())
            }),
        email: header_value(headers, "x-aimux-actor-email")
            .map(str::to_owned)
            .or_else(|| json_actor.as_ref().and_then(|actor| actor.email.clone())),
        share_id: header_value(headers, "x-aimux-share-id").map(str::to_owned),
        share_session_id: header_value(headers, "x-aimux-share-session-id").map(str::to_owned),
        principal: None,
    })
}

pub fn assert_remote_access_allowed(
    actor: Option<&RemoteActor>,
    method: &str,
    pathname: &str,
    route_url: &DaemonRouteUrl,
    context: RemoteAccessContext<'_>,
) -> RemoteAccessDecision {
    let Some(actor) = actor else {
        return RemoteAccessDecision::allow();
    };
    match actor.role {
        RemoteActorRole::Owner => RemoteAccessDecision::allow(),
        RemoteActorRole::Operator => {
            assert_operator_allowed(actor, method, pathname, route_url, context)
        }
        RemoteActorRole::Guest => assert_guest_allowed(actor, method, pathname, route_url, context),
    }
}

pub fn assert_operator_stream_allowed(
    actor: Option<&RemoteActor>,
    method: &str,
    pathname: &str,
    route_url: &DaemonRouteUrl,
    context: RemoteAccessContext<'_>,
) -> RemoteAccessDecision {
    let Some(actor) = actor else {
        return RemoteAccessDecision::deny(403, "streaming requires an operator");
    };
    if actor.role != RemoteActorRole::Operator {
        return RemoteAccessDecision::deny(403, "streaming requires an operator");
    }
    if !method.eq_ignore_ascii_case("GET") {
        return RemoteAccessDecision::deny(403, "streams are GET only");
    }

    let (principal, sub_path) = match operator_sub_path(actor, pathname) {
        Ok(value) => value,
        Err(decision) => return decision,
    };
    if sub_path != project_routes::agents::OUTPUT_STREAM {
        return RemoteAccessDecision::deny(403, "route is not available to operators");
    }
    assert_granted_session(principal, "GET", route_url, context)
}

fn assert_operator_allowed(
    actor: &RemoteActor,
    method: &str,
    pathname: &str,
    route_url: &DaemonRouteUrl,
    context: RemoteAccessContext<'_>,
) -> RemoteAccessDecision {
    let (principal, sub_path) = match operator_sub_path(actor, pathname) {
        Ok(value) => value,
        Err(decision) => return decision,
    };
    let allowed_method = operator_allowed_method(&sub_path);
    let Some(allowed_method) = allowed_method else {
        return RemoteAccessDecision::deny(403, "route is not available to operators");
    };
    if !method.eq_ignore_ascii_case(allowed_method) {
        return RemoteAccessDecision::deny(403, "method not allowed for this route");
    }
    assert_granted_session(principal, allowed_method, route_url, context)
}

fn assert_guest_allowed(
    actor: &RemoteActor,
    method: &str,
    pathname: &str,
    route_url: &DaemonRouteUrl,
    context: RemoteAccessContext<'_>,
) -> RemoteAccessDecision {
    let method_name = method.to_ascii_uppercase();
    if method_name == "GET" && pathname == "/health" {
        return RemoteAccessDecision::allow();
    }
    let Some(proxy) = parse_proxy_target(pathname) else {
        return RemoteAccessDecision::deny(403, "shared guests cannot access daemon routes");
    };
    let sub_path = proxy.sub_path;
    if method_name == "GET"
        && (shared_guest_read_route(&sub_path) || attachment_content_route(&sub_path))
    {
        return assert_shared_guest_session(actor, "GET", route_url, context);
    }
    if method_name == "POST" && shared_guest_write_route(&sub_path) {
        let session = assert_shared_guest_session(actor, "POST", route_url, context);
        if !session.ok {
            return session;
        }
        if sub_path == project_routes::live_pane::INPUT
            && !body_text(context.body)
            && !body_has_attachments(context.body)
        {
            return RemoteAccessDecision::deny(
                403,
                "shared guest input requires text or attachments",
            );
        }
        return RemoteAccessDecision::allow();
    }
    if method_name != "GET" {
        return RemoteAccessDecision::deny(
            403,
            "shared guests can only write to their shared session",
        );
    }
    RemoteAccessDecision::deny(
        403,
        "shared guests can only read shared session output and attachments",
    )
}

fn operator_sub_path<'a>(
    actor: &'a RemoteActor,
    pathname: &str,
) -> Result<(&'a RemoteOperatorPrincipal, String), RemoteAccessDecision> {
    let Some(principal) = actor.principal.as_ref() else {
        return Err(RemoteAccessDecision::deny(
            403,
            "operator actor is missing its principal",
        ));
    };
    let Some(proxy) = parse_proxy_target(pathname) else {
        return Err(RemoteAccessDecision::deny(
            403,
            "operators cannot access daemon routes",
        ));
    };
    Ok((principal, proxy.sub_path))
}

fn assert_granted_session(
    principal: &RemoteOperatorPrincipal,
    method: &str,
    route_url: &DaemonRouteUrl,
    context: RemoteAccessContext<'_>,
) -> RemoteAccessDecision {
    let Some(project_root) = context.project_root.filter(|value| !value.is_empty()) else {
        return RemoteAccessDecision::deny(403, "operator request could not be bound to a project");
    };
    let session = match authorized_session_id(method, route_url, context.body, "operator route") {
        Ok(session) => session,
        Err(error) => return RemoteAccessDecision::deny(403, error),
    };
    if !principal_has_grant(
        principal,
        &RemoteOperatorGrant {
            project_root: project_root.to_owned(),
            session_id: session,
        },
    ) {
        return RemoteAccessDecision::deny(403, "operator is not granted this session");
    }
    RemoteAccessDecision::allow()
}

fn assert_shared_guest_session(
    actor: &RemoteActor,
    method: &str,
    route_url: &DaemonRouteUrl,
    context: RemoteAccessContext<'_>,
) -> RemoteAccessDecision {
    let Some(share_session_id) = actor
        .share_session_id
        .as_deref()
        .map(str::trim)
        .filter(|value| !value.is_empty())
    else {
        return RemoteAccessDecision::deny(
            403,
            "shared guest route requires an authorized share session",
        );
    };
    let session =
        match authorized_session_id(method, route_url, context.body, "shared session route") {
            Ok(session) => session,
            Err(error) => return RemoteAccessDecision::deny(403, error),
        };
    if session != share_session_id {
        return RemoteAccessDecision::deny(403, "shared guest cannot access another session");
    }
    RemoteAccessDecision::allow()
}

fn authorized_session_id(
    method: &str,
    route_url: &DaemonRouteUrl,
    body: Option<&Value>,
    label: &str,
) -> Result<String, String> {
    let from_body = body_session_id(body);
    let from_query = route_url
        .search_param("sessionId")
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_owned);
    let method = method.to_ascii_uppercase();
    let authoritative = if method == "POST" {
        &from_body
    } else {
        &from_query
    };
    let other = if method == "POST" {
        &from_query
    } else {
        &from_body
    };
    let Some(authoritative) = authoritative.as_ref() else {
        return Err(format!("{label} requires a session id"));
    };
    if let Some(other) = other
        && other != authoritative
    {
        return Err("conflicting session ids in request body and query".into());
    }
    Ok(authoritative.clone())
}

fn principal_has_grant(principal: &RemoteOperatorPrincipal, grant: &RemoteOperatorGrant) -> bool {
    if principal.revoked_at.is_some() {
        return false;
    }
    let Some(project_root) = normalize_absolute_path(&grant.project_root) else {
        return false;
    };
    let session_id = grant.session_id.trim();
    if session_id.is_empty() {
        return false;
    }
    principal.grants.iter().any(|entry| {
        entry.project_root == project_root.to_string_lossy() && entry.session_id == session_id
    })
}

fn operator_allowed_method(sub_path: &str) -> Option<&'static str> {
    match sub_path {
        project_routes::agents::OUTPUT => Some("GET"),
        project_routes::agents::INPUT => Some("POST"),
        project_routes::agents::PROMPT_CONTEXT => Some("POST"),
        project_routes::agents::INTERRUPT => Some("POST"),
        project_routes::ATTACHMENTS => Some("POST"),
        path if attachment_content_route(path) => Some("GET"),
        _ => None,
    }
}

fn shared_guest_read_route(sub_path: &str) -> bool {
    matches!(
        sub_path,
        project_routes::agents::OUTPUT
            | project_routes::agents::HISTORY
            | project_routes::live_pane::OUTPUT
            | project_routes::EVENTS
    )
}

fn shared_guest_write_route(sub_path: &str) -> bool {
    matches!(
        sub_path,
        project_routes::live_pane::INPUT | project_routes::ATTACHMENTS
    )
}

fn attachment_content_route(path: &str) -> bool {
    let Some(rest) = path.strip_prefix("/attachments/") else {
        return false;
    };
    let Some(id) = rest.strip_suffix("/content") else {
        return false;
    };
    (1..=128).contains(&id.len())
        && id
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || byte == b'_' || byte == b'-')
}

fn body_session_id(body: Option<&Value>) -> Option<String> {
    body?
        .get("sessionId")
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_owned)
}

fn body_text(body: Option<&Value>) -> bool {
    body.and_then(|body| body.get("text"))
        .and_then(Value::as_str)
        .map(str::trim)
        .is_some_and(|value| !value.is_empty())
}

fn body_has_attachments(body: Option<&Value>) -> bool {
    body.and_then(|body| body.get("attachmentIds"))
        .and_then(Value::as_array)
        .is_some_and(|ids| {
            ids.iter().any(|id| match id {
                Value::String(value) => !value.trim().is_empty(),
                Value::Null => false,
                _ => true,
            })
        })
}

fn header_value<'a>(headers: &'a BTreeMap<String, String>, name: &str) -> Option<&'a str> {
    headers
        .iter()
        .find(|(key, value)| key.eq_ignore_ascii_case(name) && !value.trim().is_empty())
        .map(|(_, value)| value.trim())
}

fn has_relay_actor_headers(headers: &BTreeMap<String, String>) -> bool {
    headers
        .keys()
        .any(|key| key.to_ascii_lowercase().starts_with("x-aimux-"))
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct HeaderJsonActor {
    role: Option<String>,
    user_id: Option<String>,
    display_name: Option<String>,
    email: Option<String>,
}

fn actor_from_json(value: &str) -> Option<HeaderJsonActor> {
    let mut actor: HeaderJsonActor = serde_json::from_str(value).ok()?;
    if !matches!(actor.role.as_deref(), Some("owner" | "guest")) {
        actor.role = None;
    }
    Some(actor)
}

fn normalize_absolute_path(path: &str) -> Option<PathBuf> {
    let path = path.trim();
    if path.is_empty() {
        return None;
    }
    let path = Path::new(path);
    if !path.is_absolute() {
        return None;
    }
    Some(path_clean(path))
}

fn path_clean(path: &Path) -> PathBuf {
    let mut output = PathBuf::new();
    for component in path.components() {
        match component {
            std::path::Component::RootDir | std::path::Component::Prefix(_) => {
                output.push(component)
            }
            std::path::Component::CurDir => {}
            std::path::Component::ParentDir => {
                output.pop();
            }
            std::path::Component::Normal(part) => output.push(part),
        }
    }
    output
}
